File: basic_dsa_one_hundred_sixty.java
/**
 * Basic DSA Problems - Invert Binary Tree
 * Swap left and right children at every node
 */
public class basic_dsa_one_hundred_sixty {

    // Problem: Invert a binary tree (mirror it). Swap left and right of every node.

    public static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    /**
     * Recursive: invert left and right subtrees, then swap left/right of root.
     * Time O(n), Space O(h)
     */
    public static TreeNode invertTree(TreeNode root) {
        if (root == null) return null;
        TreeNode L = invertTree(root.left);
        TreeNode R = invertTree(root.right);
        root.left = R;
        root.right = L;
        return root;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Invert Binary Tree");
        System.out.println("=".repeat(60));

        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(2);
        root.right = new TreeNode(7);
        root.left.left = new TreeNode(1);
        root.left.right = new TreeNode(3);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(9);

        invertTree(root);
        System.out.println("Tree inverted (4: L=7, R=2; 7: L=9, R=6; 2: L=3, R=1)");
    }
}

File: basic_dsa_one_hundred_sixty.cpp
/**
 * Basic DSA - Symmetric Tree (LeetCode 101)
 * Check if binary tree is mirror of itself.
 */
#include <iostream>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode* l, TreeNode* r) : val(x), left(l), right(r) {}
};

bool isMirror(TreeNode* a, TreeNode* b) {
    if (!a && !b) return true;
    if (!a || !b || a->val != b->val) return false;
    return isMirror(a->left, b->right) && isMirror(a->right, b->left);
}

bool isSymmetric(TreeNode* root) {
    return !root || isMirror(root->left, root->right);
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Symmetric Tree\n";
    std::cout << std::string(60, '=') << "\n";

    TreeNode* a = new TreeNode(1,
        new TreeNode(2, new TreeNode(3), new TreeNode(4)),
        new TreeNode(2, new TreeNode(4), new TreeNode(3)));
    std::cout << "[1,2,2,3,4,4,3] -> " << (isSymmetric(a) ? "true" : "false") << "\n";

    TreeNode* b = new TreeNode(1,
        new TreeNode(2, nullptr, new TreeNode(3)),
        new TreeNode(2, nullptr, new TreeNode(3)));
    std::cout << "[1,2,2,null,3,null,3] -> " << (isSymmetric(b) ? "true" : "false") << "\n";

    std::cout << "[] -> " << (isSymmetric(nullptr) ? "true" : "false") << "\n";
}
